### PR TITLE
now trying to fit previously selected cell into newly selected item #71

### PR DESCRIPTION
**Pull request by IvanKarpan**
**What I did:**
Added logic to check if a newly selected item can reuse previously selected cells

**Why I did it:**
This feature was requested

**Fixes issue:**
https://github.com/atomicptr/dauntless-builder/issues/71

### CODE REVIEW

**Comment by atomicptr** [comment]
One issue I've found is with items which have the two of the same cell type like Infernos Razor:

image

if I now switch to something else like Charred Saber it keeps the first +1, but in data it stays at an invalid state

image

Another example would be:

image

image

**Comment by IvanKarpan** [comment]
Oh, I didn't even know there are weapons like this so have not tested that, I'll take a look and update the PR, thanks for spotting it

**Comment by IvanKarpan** [comment]
hey, rewrote the logic, checked the cases you've mentioned and it seems to handle them... seems means it could use double checking, which something tells me you'd do in either case (thank you for this)

**Comment by atomicptr** [comment]
The site just crashes for me right now :).

## Steps to reproduce

1. Go to https://deploy-preview-100--dauntless-builder.netlify.com/b/new
2. Select a weapon

**Comment by IvanKarpan** [comment]
Fixed, saw an opportunity to optimize the code right before I committed it and moved a variables declaration before its first usage... :-/

**Comment by atomicptr** [comment]
I found another edge case :)

Should a cell move from slot 1 to 2 if 2 matches the slot? :thinking: 

Drask War Pike:

image

Drask Sword:

image

Also switching:

image

image

The question is if this should work :thinking:

**Comment by IvanKarpan** [comment]
It definitely should and I kind of thought I actually handled this case... Thanks for the QA, I'll check why it is not working.

**Comment by IvanKarpan** [comment]
Well, I seem to have gotten myself completely confused about this logic the other day... Rewrote it and - I know I said this before - but I really think it should work properly this time :D

File: src/routes/BuildRoute.jsx [hunk 1]
```diff
@@ -157,18 +155,24 @@ export default class BuildRoute extends React.Component {
...
             
-            const cellKeys = [
+            const changesKeys = [
                 "weapon_cell0",
                 "weapon_cell1"
             ];
-            const cells = this.getCellsForKeys(cellKeys);
-            if (cells.length && item.cells) {
+            const cells = this.getCellsForKeys(changesKeys);
+            console.log("cells", cells);
```
Comment (atomicptr) [inline src/routes/BuildRoute.jsx]: You forgot something :see_no_evil:

File: src/routes/BuildRoute.jsx [hunk 2]
```diff
@@ -157,18 +155,24 @@ export default class BuildRoute extends React.Component {
...
+            if (Object.keys(cells).length && item.cells) {
                 item.cells.forEach((slot, index) => {
-                    const cellKey = cellKeys[index];
-                    const cellIndex = cells.findIndex(cell => cell.slot === slot);
-                    if (cellIndex !== -1) {
-                        changes[cellKey] = this.state.build[cellKey];
-                        cells.splice(cellIndex, 1);
+                    console.log("slot", slot);
```
Comment (atomicptr) [inline src/routes/BuildRoute.jsx]: Another one :D

File: src/routes/BuildRoute.jsx [hunk 3]
```diff
@@ -157,18 +155,24 @@ export default class BuildRoute extends React.Component {
...
+                    const changesKey = changesKeys[index];
+                    for (const cellKey in cells) {
+                        console.log("cells[cellKey]", cells[cellKey]);
```
Comment (atomicptr) [inline src/routes/BuildRoute.jsx]: And another one

File: src/routes/BuildRoute.jsx [hunk 4]
```diff
@@ -157,18 +155,24 @@ export default class BuildRoute extends React.Component {
...
+                        if (cells[cellKey] && cells[cellKey].slot === slot) {
+                            console.log("this.state.build[cellKey]", this.state.build[cellKey]);
```
Comment (atomicptr) [inline src/routes/BuildRoute.jsx]: Last one I comment at least :)

**Comment by atomicptr** [comment]
This is a new behaviour, but I think it's fine-ish

Going from:

image

To:

image

Should probably use the same cell on the same slot instead of moving the other one to slot 2? Dunno, weird edge case.

**Comment by atomicptr** [comment]
Tho, I really want to merge this, the QoL this brings is huge :)

**Comment by IvanKarpan** [comment]
Cleaned up the logs, as for the last case... Well this is a matter of preference I believe - some people would want one behavior, some the other. I vote for not complicating the logic any further)))

**Comment by atomicptr** [comment]
I guess thats true, I also doubt most people will notice :D

**Comment by IvanKarpan** [comment]
It depends on how you look at this - arguably you would've put the cell that's more important to the build in the first slot, and secondary in the second.

**Comment by atomicptr** [comment]
Oh, merging #102 caused a conflict :see_no_evil:. If you don't want to, I'll resolve that later when I get home. Beyond that I'll merge this as is.

**Comment by IvanKarpan** [comment]
Resolved the conflict, no problem

**Pull request merged**